# WEEK 6 (200968038), SHIV PRATAP SINGH

#### Q1) Read about the movielens dataset and write down a summary of metadata.

MOVIESLENS DATASET
- (whole dataset is in the form of comma seperated values)
    -The data set consists of 4 csv files named links, movies,ratings and tags
        - ratings.csv - all ratings contained in the file ratings.csv
                      - ratings made from 0-5 with 0.5 incrementation
        - tags.csv - contains all the tags
        - movies.csv - contains information about movie titles and the gneres associated with the movie
        - links.csv - contains identifiers that can be used to link other sources of movies data
                    - movieId - identifier movieLens
                    - imdbId - identifier imdb 
                    - tmdbId - identifier  theMovieDb
    -This data set describes a 5 - star rating and free-text tagging activity from MovieLens(a movie recommender service) 
    - 100836 ratings
    - 3683 tag applications
    - 9742 movies
    - users involed in creation = 610
    - duration of collection of data - March 29, 1996 - September 24, 2018
    - every user rated atleaset 20 ratings

#### Q2) Read the “ratings.csv” file and create a pivot table with index=‘userId’, columns=‘movieId’,values = “rating. 

In [150]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [151]:
ratings = pd.read_csv('ratings.csv')
ratings

userId  movieId  rating   timestamp
0            1        1     4.0   964982703
1            1        3     4.0   964981247
2            1        6     4.0   964982224
3            1       47     5.0   964983815
4            1       50     5.0   964982931
...        ...      ...     ...         ...
100831     610   166534     4.0  1493848402
100832     610   168248     5.0  1493850091
100833     610   168250     5.0  1494273047
100834     610   168252     5.0  1493846352
100835     610   170875     3.0  1493846415

[100836 rows x 4 columns]

In [152]:
ratings.isnull().sum()

userId       0
movieId      0
rating       0
timestamp    0
dtype: int64

In [153]:
 ratings.movieId.max()

193609

In [156]:
rating_pivot = ratings.pivot_table(index='userId',columns='movieId',values='rating',fill_value=0)
rating_pivot# Now we have a sparse matrix representation

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           4.0     0.0     4.0     0.0     0.0     4.0     0.0       0   
2           0.0     0.0     0.0     0.0     0.0     0.0     0.0       0   
3           0.0     0.0     0.0     0.0     0.0     0.0     0.0       0   
4           0.0     0.0     0.0     0.0     0.0     0.0     0.0       0   
5           4.0     0.0     0.0     0.0     0.0     0.0     0.0       0   
...         ...     ...     ...     ...     ...     ...     ...     ...   
606         2.5     0.0     0.0     0.0     0.0     0.0     2.5       0   
607         4.0     0.0     0.0     0.0     0.0     0.0     0.0       0   
608         2.5     2.0     2.0     0.0     0.0     0.0     0.0       0   
609         3.0     0.0     0.0     0.0     0.0     0.0     0.0       0   
610         5.0     0.0     0.0     0.0     0.0     5.0     0.0       0   

movieId  9       10      ...  193565  193567  193571  193573  193579  193581  \
userId                   ...                                                   
1           0.0     0.0  ...     0.0       0       0       0     0.0       0   
2           0.0     0.0  ...     0.0       0       0       0     0.0       0   
3           0.0     0.0  ...     0.0       0       0       0     0.0       0   
4           0.0     0.0  ...     0.0       0       0       0     0.0       0   
5           0.0     0.0  ...     0.0       0       0       0     0.0       0   
...         ...     ...  ...     ...     ...     ...     ...     ...     ...   
606         0.0     0.0  ...     0.0       0       0       0     0.0       0   
607         0.0     0.0  ...     0.0       0       0       0     0.0       0   
608         0.0     4.0  ...     0.0       0       0       0     0.0       0   
609         0.0     4.0  ...     0.0       0       0       0     0.0       0   
610         0.0     0.0  ...     0.0       0       0       0     0.0       0   

movieId  193583  193585  193587  193609  
userId                                   
1           0.0     0.0     0.0       0  
2           0.0     0.0     0.0       0  
3           0.0     0.0     0.0       0  
4           0.0     0.0     0.0       0  
5           0.0     0.0     0.0       0  
...         ...     ...     ...     ...  
606         0.0     0.0     0.0       0  
607         0.0     0.0     0.0       0  
608         0.0     0.0     0.0       0  
609         0.0     0.0     0.0       0  
610         0.0     0.0     0.0       0  

[610 rows x 9724 columns]

#### Q3) sklearn.metrics.pairwise_distances can be used to compute distance between all pairs of users. pairwise_distances() takes a metric parameter for what distance measure to use. Use cosine similarity for finding similarity among users. Use the following packages.

from sklearn.metrics import pairwise_distances

from scipy.spatial.distance import cosine, correlation

In [8]:
from sklearn.metrics import pairwise_distances
from scipy.spatial.distance import cosine, correlation

In [160]:
distances = pairwise_distances(rating_pivot,metric='cosine')
distances

array([[0.00000000e+00, 9.72717135e-01, 9.40279738e-01, ...,
        7.08902628e-01, 9.06428070e-01, 8.54679193e-01],
       [9.72717135e-01, 1.11022302e-16, 1.00000000e+00, ...,
        9.53789046e-01, 9.72434599e-01, 8.97573246e-01],
       [9.40279738e-01, 1.00000000e+00, 0.00000000e+00, ...,
        9.78871538e-01, 1.00000000e+00, 9.67881252e-01],
       ...,
       [7.08902628e-01, 9.53789046e-01, 9.78871538e-01, ...,
        0.00000000e+00, 8.78007286e-01, 6.77945142e-01],
       [9.06428070e-01, 9.72434599e-01, 1.00000000e+00, ...,
        8.78007286e-01, 0.00000000e+00, 9.46774537e-01],
       [8.54679193e-01, 8.97573246e-01, 9.67881252e-01, ...,
        6.77945142e-01, 9.46774537e-01, 1.11022302e-16]])

#### Q 6. Find the 5 most similar user for user with user Id 10.

In [163]:
(-distances[9]).argsort()[1:6]+1

array([194, 521, 206, 214,  90], dtype=int64)

We select the 9th row from the resultant distance matrix to select the 10th User. (Arrays in Python are 0-indexed and the UserIDs start from 1)

Then, we sort the cosine similarities in the descending order to find the smallest distance between users, and hence the most similar users. We use the argsort() function on the negated array to return the indices, and hence the UserIDs of the sorted array. Then, we splice the array to find the top 5 similar users. We leave out the index 0 because it is by default the same user. 

After splicing, we add 1 to the indices since arrays in Python are 0-indexed and UserIDs start from 1.

#### This shows that users 194, 521, 206, 214, and 90 are most similar to user 10 (This is 0-based system in the distance matrix)

#### Q7) Use the “movies” dataset to find out the names of movies, user 2 and user 338 have watched in common and how they have rated each one of them.


In [164]:
movies = pd.read_csv('movies.csv')
movies

movieId                                      title  \
0           1                           Toy Story (1995)   
1           2                             Jumanji (1995)   
2           3                    Grumpier Old Men (1995)   
3           4                   Waiting to Exhale (1995)   
4           5         Father of the Bride Part II (1995)   
...       ...                                        ...   
9737   193581  Black Butler: Book of the Atlantic (2017)   
9738   193583               No Game No Life: Zero (2017)   
9739   193585                               Flint (2017)   
9740   193587        Bungo Stray Dogs: Dead Apple (2018)   
9741   193609        Andrew Dice Clay: Dice Rules (1991)   

                                           genres  
0     Adventure|Animation|Children|Comedy|Fantasy  
1                      Adventure|Children|Fantasy  
2                                  Comedy|Romance  
3                            Comedy|Drama|Romance  
4                                          Comedy  
...                                           ...  
9737              Action|Animation|Comedy|Fantasy  
9738                     Animation|Comedy|Fantasy  
9739                                        Drama  
9740                             Action|Animation  
9741                                       Comedy  

[9742 rows x 3 columns]

This dataframe can use some data cleaning

In [165]:
movies.drop('genres',axis=1,inplace=True)

In [166]:
# Separting movie title and year part using split function.
split_values = movies['title'].str.split("(", n = 1, expand = True)
# setting 'movie_title' values to title part.
movies.title = split_values[0]
# creating 'release_year' column.
movies['release_year'] = split_values[1]
# Cleaning the release_year series.
movies['release_year'] = movies.release_year.str.replace(')','')


In [167]:
movies

movieId                                title release_year
0           1                           Toy Story          1995
1           2                             Jumanji          1995
2           3                    Grumpier Old Men          1995
3           4                   Waiting to Exhale          1995
4           5         Father of the Bride Part II          1995
...       ...                                  ...          ...
9737   193581  Black Butler: Book of the Atlantic          2017
9738   193583               No Game No Life: Zero          2017
9739   193585                               Flint          2017
9740   193587        Bungo Stray Dogs: Dead Apple          2018
9741   193609        Andrew Dice Clay: Dice Rules          1991

[9742 rows x 3 columns]

In [92]:
merge = pd.merge(ratings,movies,on='movieId')
merge

userId  movieId  rating   timestamp                       title  \
0            1        1     4.0   964982703                  Toy Story    
1            5        1     4.0   847434962                  Toy Story    
2            7        1     4.5  1106635946                  Toy Story    
3           15        1     2.5  1510577970                  Toy Story    
4           17        1     4.5  1305696483                  Toy Story    
...        ...      ...     ...         ...                         ...   
100831     610   160341     2.5  1479545749                  Bloodmoon    
100832     610   160527     4.5  1479544998  Sympathy for the Underdog    
100833     610   160836     3.0  1493844794                     Hazard    
100834     610   163937     3.5  1493848789                Blair Witch    
100835     610   163981     3.5  1493850155                         31    

       release_year  
0              1995  
1              1995  
2              1995  
3              1995  
4              1995  
...             ...  
100831         1997  
100832         1971  
100833         2005  
100834         2016  
100835         2016  

[100836 rows x 6 columns]

In [93]:
merge[(merge.userId==2) | (merge.userId==338)].movieId

692          50
1997        296
2267        333
3684        527
4462        593
          ...  
97114    190209
97115    190213
97116    190215
97117    190219
97118    190221
Name: movieId, Length: 68, dtype: int64

In [22]:
import numpy as np
c1 = merge['title'][merge.userId==2]
c2 = merge['title'][merge.userId==338]
common = np.intersect1d(c1,c2)
common

array(['Kill Bill: Vol. 1 ', 'Shawshank Redemption, The '], dtype=object)

These are the movies that users 2 and 338 have watched

In [23]:
cond=((merge.title=='Kill Bill: Vol. 1 ') | (merge.title=='Shawshank Redemption, The ')) & ((merge.userId==2) | (merge.userId==338))
Q7 = merge[cond]
Q7

userId  movieId  rating   timestamp                       title  IMAX  \
16296       2      318     3.0  1445714835  Shawshank Redemption, The      0   
16473     338      318     5.0  1530148309  Shawshank Redemption, The      0   
16754       2     6874     4.0  1445714952          Kill Bill: Vol. 1      0   
16831     338     6874     4.5  1530148453          Kill Bill: Vol. 1      0   

       Musical  Fantasy  Western  Adventure  ...  (no genres listed)  \
16296        0        0        0          0  ...                   0   
16473        0        0        0          0  ...                   0   
16754        0        0        0          0  ...                   0   
16831        0        0        0          0  ...                   0   

       Film-Noir  Documentary  Sci-Fi  War  Thriller  Animation  Crime  \
16296          0            0       0    0         0          0      1   
16473          0            0       0    0         0          0      1   
16754          0            0       0    0         1          0      1   
16831          0            0       0    0         1          0      1   

       Action  release_year  
16296       0          1994  
16473       0          1994  
16754       1          2003  
16831       1          2003  

[4 rows x 26 columns]

In [24]:
Q7 = Q7.pivot(index='title',columns='userId',values='rating')
Q7

userId                      2    338
title                               
Kill Bill: Vol. 1           4.0  4.5
Shawshank Redemption, The   3.0  5.0

This is how the users 2 and 338 have rated the movies they've watched together

#### Q8) Use the movies dataset to find out the common movie names between user 2 and user 338 with least rating of 4.0

In [25]:
Q7[(Q7[2]>=4) & (Q7[338]>=4)]

userId              2    338
title                       
Kill Bill: Vol. 1   4.0  4.5

Kill Bill: Vol 1 is the movie which the users 2 and 338 have given min 4 rating

#### Q9) Create a pivot table for representing the similarity among movies using correlation.

In [168]:
movie_pivot = ratings.pivot_table(index='movieId',columns='userId',values='rating',fill_value=0)
movie_pivot

userId   1    2    3    4    5    6    7    8    9    10   ...  601  602  603  \
movieId                                                    ...                  
1          4  0.0  0.0    0    4    0  4.5    0    0  0.0  ...  4.0    0    4   
2          0  0.0  0.0    0    0    4  0.0    4    0  0.0  ...  0.0    4    0   
3          4  0.0  0.0    0    0    5  0.0    0    0  0.0  ...  0.0    0    0   
4          0  0.0  0.0    0    0    3  0.0    0    0  0.0  ...  0.0    0    0   
5          0  0.0  0.0    0    0    5  0.0    0    0  0.0  ...  0.0    0    0   
...      ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
193581     0  0.0  0.0    0    0    0  0.0    0    0  0.0  ...  0.0    0    0   
193583     0  0.0  0.0    0    0    0  0.0    0    0  0.0  ...  0.0    0    0   
193585     0  0.0  0.0    0    0    0  0.0    0    0  0.0  ...  0.0    0    0   
193587     0  0.0  0.0    0    0    0  0.0    0    0  0.0  ...  0.0    0    0   
193609     0  0.0  0.0    0    0    0  0.0    0    0  0.0  ...  0.0    0    0   

userId   604  605  606  607  608  609  610  
movieId                                     
1          3  4.0  2.5    4  2.5    3  5.0  
2          5  3.5  0.0    0  2.0    0  0.0  
3          0  0.0  0.0    0  2.0    0  0.0  
4          0  0.0  0.0    0  0.0    0  0.0  
5          3  0.0  0.0    0  0.0    0  0.0  
...      ...  ...  ...  ...  ...  ...  ...  
193581     0  0.0  0.0    0  0.0    0  0.0  
193583     0  0.0  0.0    0  0.0    0  0.0  
193585     0  0.0  0.0    0  0.0    0  0.0  
193587     0  0.0  0.0    0  0.0    0  0.0  
193609     0  0.0  0.0    0  0.0    0  0.0  

[9724 rows x 610 columns]

In [169]:
distances = pd.DataFrame(pairwise_distances(movie_pivot, metric='correlation'))
distances

0         1         2         3         4         5         6     \
0     0.000000  0.768673  0.826787  1.028917  0.807526  0.807314  0.856257   
1     0.768673  0.000000  0.808055  0.928731  0.799474  0.841659  0.872431   
2     0.826787  0.808055  0.000000  0.932857  0.629829  0.803558  0.648487   
3     1.028917  0.928731  0.932857  0.000000  0.832090  0.946245  0.741925   
4     0.807526  0.799474  0.629829  0.832090  0.000000  0.784497  0.570110   
...        ...       ...       ...       ...       ...       ...       ...   
9719  1.028906  1.018291  1.011729  1.004138  1.011456  1.017712  1.012033   
9720  1.028906  1.018291  1.011729  1.004138  1.011456  1.017712  1.012033   
9721  1.028906  1.018291  1.011729  1.004138  1.011456  1.017712  1.012033   
9722  1.028906  1.018291  1.011729  1.004138  1.011456  1.017712  1.012033   
9723  1.028906  1.018291  1.011729  1.004138  1.011456  1.017712  1.012033   

          7         8         9     ...          9714          9715  \
0     0.914523  0.822755  0.816618  ...  1.028906e+00  1.028906e+00   
1     0.858460  1.021045  0.714914  ...  1.018291e+00  1.018291e+00   
2     0.703103  0.724188  0.863084  ...  1.011729e+00  1.011729e+00   
3     0.851274  1.016025  0.944000  ...  1.004138e+00  1.004138e+00   
4     0.734223  0.691915  0.889167  ...  1.011456e+00  1.011456e+00   
...        ...       ...       ...  ...           ...           ...   
9719  1.004383  1.006359  1.020524  ...  2.275957e-14  3.941292e-14   
9720  1.004383  1.006359  1.020524  ...  0.000000e+00  2.831069e-14   
9721  1.004383  1.006359  1.020524  ...  0.000000e+00  2.831069e-14   
9722  1.004383  1.006359  1.020524  ...  0.000000e+00  2.831069e-14   
9723  1.004383  1.006359  0.939141  ...  1.001642e+00  1.001642e+00   

              9716          9717          9718          9719          9720  \
0     1.028906e+00  1.028906e+00  1.028906e+00  1.028906e+00  1.028906e+00   
1     1.018291e+00  1.018291e+00  1.018291e+00  1.018291e+00  1.018291e+00   
2     1.011729e+00  1.011729e+00  1.011729e+00  1.011729e+00  1.011729e+00   
3     1.004138e+00  1.004138e+00  1.004138e+00  1.004138e+00  1.004138e+00   
4     1.011456e+00  1.011456e+00  1.011456e+00  1.011456e+00  1.011456e+00   
...            ...           ...           ...           ...           ...   
9719  0.000000e+00  0.000000e+00  2.275957e-14  0.000000e+00  2.275957e-14   
9720  2.275957e-14  2.275957e-14  0.000000e+00  2.275957e-14  0.000000e+00   
9721  2.275957e-14  2.275957e-14  0.000000e+00  2.275957e-14  0.000000e+00   
9722  2.275957e-14  2.275957e-14  0.000000e+00  2.275957e-14  0.000000e+00   
9723  1.001642e+00  1.001642e+00  1.001642e+00  1.001642e+00  1.001642e+00   

              9721          9722      9723  
0     1.028906e+00  1.028906e+00  1.028906  
1     1.018291e+00  1.018291e+00  1.018291  
2     1.011729e+00  1.011729e+00  1.011729  
3     1.004138e+00  1.004138e+00  1.004138  
4     1.011456e+00  1.011456e+00  1.011456  
...            ...           ...       ...  
9719  2.275957e-14  2.275957e-14  1.001642  
9720  0.000000e+00  0.000000e+00  1.001642  
9721  0.000000e+00  0.000000e+00  1.001642  
9722  0.000000e+00  0.000000e+00  1.001642  
9723  1.001642e+00  1.001642e+00  0.000000  

[9724 rows x 9724 columns]

Here's the pivot table for representing the similarity among movies

#### Q10) Find the top 5  movies which are similar to the movie “Godfather”(MovieId = 858)

In [170]:
movies[movies.movieId==858]#The index number of the MovieId 858 is 659

movieId            title release_year
659      858  Godfather, The          1972

In [172]:
for i in distances.iloc[659].nsmallest(6).index.tolist()[1:]:
    print(movies.title.iloc[i])

Blues Brothers, The 
Third Man, The 
Paris Is Burning 
Dirty Dancing 
Fargo 


These are the 5 movies similar to Godfather